sql/src/main/java/org/apache/druid/sql/calcite/rel/VirtualColumnRegistry.java

```diff
@@ -169,15 +212,75 @@ public RowSignature getFullRowSignature()
                      .collect(Collectors.toList());
   }
 
-  private static class ExpressionWrapper
+  /**
+   * @deprecated use {@link #getOrCreateVirtualColumnForExpression(DruidExpression, ColumnType)} instead
+   */
+  @Deprecated
+  public VirtualColumn getOrCreateVirtualColumnForExpression(
+      PlannerContext plannerContext,
+      DruidExpression expression,
+      ColumnType valueType
+  )
+  {
+    final String name = getOrCreateVirtualColumnForExpression(expression, valueType);
+    return virtualColumnsByName.get(name).expression.toVirtualColumn(name, valueType, macroTable);
+  }
+
+  /**
+   * @deprecated use {@link #getOrCreateVirtualColumnForExpression(DruidExpression, RelDataType)} instead
+   */
+  @Deprecated
+  public VirtualColumn getOrCreateVirtualColumnForExpression(
+      PlannerContext plannerContext,
+      DruidExpression expression,
+      RelDataType dataType
+  )
   {
-    private final String expression;
-    private final ColumnType valueType;
+    return getOrCreateVirtualColumnForExpression(
+        plannerContext,
+        expression,
+        Calcites.getColumnTypeForRelDataType(dataType)
+    );
+  }
 
-    public ExpressionWrapper(String expression, ColumnType valueType)
+  /**
+   * @deprecated use {@link #getVirtualColumnByExpression(DruidExpression, RelDataType)} instead
+   */
+  @Deprecated
+  @Nullable
+  public VirtualColumn getVirtualColumnByExpression(String expression, RelDataType type)
+  {
+    final ColumnType columnType = Calcites.getColumnTypeForRelDataType(type);
+    ExpressionAndTypeHint wrapped = wrap(DruidExpression.fromExpression(expression), columnType);
+    return Optional.ofNullable(virtualColumnsByExpression.get(wrapped))
+                   .map(this::getVirtualColumn)
+                   .orElse(null);
+  }
+
+  private static ExpressionAndTypeHint wrap(DruidExpression expression, ColumnType typeHint)
+  {
+    return new ExpressionAndTypeHint(expression, typeHint);
+  }
+
+  private static class ExpressionAndTypeHint
+  {
+    private final DruidExpression expression;
+    private final ColumnType typeHint;
```

Comment:
nit: `typeHint` seems worth to document. Can you please add some javadoc for it? It's not a blocker and you can do it as a follow-up.